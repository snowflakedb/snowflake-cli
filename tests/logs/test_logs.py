from unittest import mock

import pytest


def test_providing_to_time_and_refresh_causes_error(runner):

    result = runner.invoke(
        [
            "logs",
            "table",
            "test_table",
            "--to",
            '"2022-02-02 02:02:02"',
            "--refresh",
            "5",
        ]
    )
    assert result.exit_code == 1
    assert "You cannot set both --refresh and --to parameters" in result.output


@mock.patch("snowflake.cli._plugins.logs.manager.LogsManager.logs_table")
def test_providing_to_time_earlier_than_from_time_causes_error(mock_table, runner):
    mock_table.return_value = "test_table"

    result = runner.invoke(
        [
            "logs",
            "table",
            "test_table",
            "--from",
            "2022-02-02 02:02:02",
            "--to",
            "2022-02-01 02:02:02",
        ]
    )
    assert result.exit_code == 1
    assert "From_time cannot be later than to_time" in result.output


@pytest.mark.parametrize("parameter", ["--from", "--to"])
@pytest.mark.parametrize(
    "time_string",
    [
        "2024-11-03 12:00:00 UTC",
        "2024.11.03 12 00",
        "About one hour ago",
        "92348573948753202",
    ],
)
def test_providing_time_in_incorrect_format_causes_error(
    time_string, parameter, runner, snapshot
):
    result = runner.invoke(["logs", "compute_pool", "foo", parameter, time_string])

    assert result.exit_code == 1
    assert result.output == snapshot


@mock.patch(
    "snowflake.cli._plugins.logs.manager.LogsManager.logs_table",
    new_callable=mock.PropertyMock,
)
def test_correct_query_is_constructed(
    mock_table, mock_connect, mock_ctx, runner, snapshot
):
    mock_table.return_value = "foo"
    ctx = mock_ctx()
    mock_connect.return_value = ctx

    _ = runner.invoke(
        [
            "logs",
            "compute_pool",
            "bar",
            "--from",
            "2022-02-02 02:02:02",
            "--to",
            "2022-02-03 02:02:02",
        ]
    )

    queries = ctx.get_queries()
    assert len(queries) == 1
    assert queries[0] == snapshot
