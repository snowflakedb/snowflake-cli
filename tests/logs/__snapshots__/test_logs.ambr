# serializer version: 1
# name: test_correct_query_is_constructed
  '''
  SELECT
                  timestamp,
                  resource_attributes:"snow.database.name"::string as database_name,
                  resource_attributes:"snow.schema.name"::string as schema_name,
                  resource_attributes:"snow.compute_pool.name"::string as object_name,
                  record:severity_text::string as log_level,
                  value::string as log_message
              FROM foo
              WHERE record_type = 'LOG'
              AND (record:severity_text = 'INFO' or record:severity_text is NULL )
              AND object_name = 'bar'
              AND timestamp >= TO_TIMESTAMP_LTZ('2022-02-02T02:02:02')
  AND timestamp <= TO_TIMESTAMP_LTZ('2022-02-03T02:02:02')
  
              ORDER BY timestamp;
  '''
# ---
# name: test_providing_time_in_incorrect_format_causes_error[2024-11-03 12:00:00 UTC---from]
  '''
  +- Error ----------------------------------------------------------------------+
  | Incorrect format for '--from'. Please use one of approved ISO formats.       |
  +------------------------------------------------------------------------------+
  
  '''
# ---
# name: test_providing_time_in_incorrect_format_causes_error[2024-11-03 12:00:00 UTC---to]
  '''
  +- Error ----------------------------------------------------------------------+
  | Incorrect format for '--to'. Please use one of approved ISO formats.         |
  +------------------------------------------------------------------------------+
  
  '''
# ---
# name: test_providing_time_in_incorrect_format_causes_error[2024.11.03 12 00---from]
  '''
  +- Error ----------------------------------------------------------------------+
  | Incorrect format for '--from'. Please use one of approved ISO formats.       |
  +------------------------------------------------------------------------------+
  
  '''
# ---
# name: test_providing_time_in_incorrect_format_causes_error[2024.11.03 12 00---to]
  '''
  +- Error ----------------------------------------------------------------------+
  | Incorrect format for '--to'. Please use one of approved ISO formats.         |
  +------------------------------------------------------------------------------+
  
  '''
# ---
# name: test_providing_time_in_incorrect_format_causes_error[92348573948753202---from]
  '''
  +- Error ----------------------------------------------------------------------+
  | Incorrect format for '--from'. Please use one of approved ISO formats.       |
  +------------------------------------------------------------------------------+
  
  '''
# ---
# name: test_providing_time_in_incorrect_format_causes_error[92348573948753202---to]
  '''
  +- Error ----------------------------------------------------------------------+
  | Incorrect format for '--to'. Please use one of approved ISO formats.         |
  +------------------------------------------------------------------------------+
  
  '''
# ---
# name: test_providing_time_in_incorrect_format_causes_error[About one hour ago---from]
  '''
  +- Error ----------------------------------------------------------------------+
  | Incorrect format for '--from'. Please use one of approved ISO formats.       |
  +------------------------------------------------------------------------------+
  
  '''
# ---
# name: test_providing_time_in_incorrect_format_causes_error[About one hour ago---to]
  '''
  +- Error ----------------------------------------------------------------------+
  | Incorrect format for '--to'. Please use one of approved ISO formats.         |
  +------------------------------------------------------------------------------+
  
  '''
# ---
